src/Abp.NHibernate/NHibernate/Repositories/NhRepositoryBaseOfTEntityAndTPrimaryKey.cs

```diff
@@ -57,11 +58,36 @@ public override IQueryable<TEntity> GetAllIncluding(params Expression<Func<TEnti
             return query;
         }
 
+        public override Task<List<TEntity>> GetAllListAsync()
+        {
+            return GetAll().ToListAsync();
+        }
+
+        public override Task<List<TEntity>> GetAllListAsync(Expression<Func<TEntity, bool>> predicate)
+        {
+            return GetAll().Where(predicate).ToListAsync();
+        }
+
+        public override Task<TEntity> SingleAsync(Expression<Func<TEntity, bool>> predicate)
+        {
+            return GetAll().SingleAsync(predicate);
+        }
+
         public override TEntity FirstOrDefault(TPrimaryKey id)
         {
             return Session.Get<TEntity>(id);
         }
 
+        public override Task<TEntity> FirstOrDefaultAsync(TPrimaryKey id)
+        {
+            return Session.GetAsync<TEntity>(id);
```

Comment:
> Any reason this is not `Session.FirstOrDefaultAsync`?

I can't find `Session.FirstOrDefaultAsync` method. There is an `IQueryable` extension method `FirstOrDefaultAsync` that takes and Expression as parameter. And we can't use `GetAll().FirstOrDefaultAsync(entity => entity.Id == id)` because Id is a generic type (TPrimaryKey) with no constraint.

There's a workaround on the base class `AbpRepositoryBase.CreateEqualityExpressionForId(id)`, but GetAsync just do all this.

> I assume that `GetAsync(I'd)` will throw exception when not found?
> 
> The convention for FirstOrDefaultAsync is not to throw exception.

No, `Session.Get(object id)` and `Session.GetAsync(object id)` have the same behavior. They return null if not found. From NHibernate documentation :

> Return the persistent instance of the given named entity with the given identifier, or null if there is no such persistent instance.

I will add later a unit test that do the call on a not found item.

Reply:
Added unit tests